File: CMPS 101 - C:Java/pa4/List.c
//  Nichole Boothroyd    cruzID: nboothro  Assignment: pa2
//-----------------------------------------------------------------------------
// List.c
// Implementation file for List ADT
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj{
    int data;
    struct NodeObj* next;
    struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
    Node front;
    Node back;
    int length;
    int index;
} ListObj;


// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
    Node N = malloc(sizeof(NodeObj));
    N->data = data;
    N->next = NULL;
    N->prev = NULL;
    return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
    if( pN!=NULL && *pN!=NULL ){
        free(*pN);
        *pN = NULL;
    }
}

// newList()
// Returns reference to new empty List object.
List newList(void){
    List L;
    L = malloc(sizeof(ListObj));
    L->front = L->back = NULL;
    L->length = 0;
    L->index = -1;
    return(L);
}


// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL.S
void freeList(List* pL){
    if(pL!=NULL && *pL!=NULL) {
        while( !isEmpty(*pL) ) {
            deleteFront(*pL);
        }
        free(*pL);
        *pL = NULL;
    }
}


// Access functions -----------------------------------------------------------

// isEmpty()
// Returns true (1) if L is empty, otherwise returns false (0)
int isEmpty(List L){
    if( L==NULL ){
        printf("List Error: calling isEmpty() on NULL List reference\n");
        exit(1);
    }
    return(L->length==0);
}

// length()
// Returns the length of L.
int length(List L){
    if( L==NULL ){
        printf("List Error: calling length() on NULL List reference\n");
        exit(1);
    }
    return(L->length);
}

//index()
//If the cursor is defined, returns the index of the cursor element
//Otherwise, return -1
int index(List L){
    if( L==NULL ){
        printf("List Error: calling index() on NULL List reference\n");
        exit(1);
    }
    return(L->index);  //cursor defaulted at -1
}

// front()
// Returns the value at the front of L.
// Pre: !isEmpty(L)
int front(List L){
    if( L==NULL ){
        printf("List Error: calling front() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling front() on an empty List\n");
        exit(1);
    }
    return(L->front->data);
}

// back()
// Returns the value at the back of L.
// Pre: !isEmpty(L)
int back(List L){
    if( L==NULL ){
        printf("List Error: calling front() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling back() on an empty List\n");
        exit(1);
    }
    return(L->back->data);
}

//get()
//Returns cursor element
//Pre: !this.isEmpty()
//Pre: index needs to be defined
int get(List L) {
    if( L==NULL ){
        printf("List Error: calling get() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling get() on an empty List\n");
        exit(1);
    }
    if( index(L) < 0) {
        printf("List Error: calling get() on an undefined cursor\n");
        exit(1);
    }
    else {
        Node temp = L->front;
        for (int i = 1; i<= index(L); i++) {
            temp = temp->next;
        }
        return(temp->data);
    }
}

// equals()
// returns true (1) if A is identical to B, false (0) otherwise
int equals(List A, List B){
    int eq = 0;
    Node N = NULL;
    Node M = NULL;
    
    if( A==NULL || B==NULL ){
        printf("List Error: calling equals() on NULL List reference\n");
        exit(1);
    }
    
    eq = ( A->length == B->length );
    N = A->front;
    M = B->front;
    while( eq && N!=NULL){
        eq = (N->data==M->data);
        N = N->next;
        M = M->next;
    }
    return eq;
}

// Manipulation procedures ----------------------------------------------------

//clear ()
//Resets this List to its original empty state
//Pre: !this.isEmpty()
void clear(List L) {
    if( L==NULL ){
        printf("List Error: calling clear() on NULL List reference\n");
        exit(1);
    }
/*    if( isEmpty(L) ){
        printf("List Error: calling clear() on an empty List\n");
        exit(1);
    }
 */
    else {
        while (length(L) > 0) {
            deleteBack(L);
        }
    }
}
    
//moveFront()
//Places the cursor under front element, or does nothing
//Pre: !this.isEmpty()
void moveFront(List L) {
    if( L==NULL ){
        printf("List Error: calling moveFront() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling moveFront() on an empty List\n");
        exit(1);
    }
    else {
        L->index = 0;
    }
}

//moveBack()
//Places the cursor under back element, or does nothing
//Pre: !this.isEmpty()
void moveBack(List L) {
    if( L==NULL ){
        printf("List Error: calling moveBack() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling moveBack() on an empty List\n");
        exit(1);
    }
    else {
        L->index = (length(L) - 1);
    }
}

//movePrev ()
//If cursor is defined and not at front, moves cursor one step toward
//front of this List, if cursor is defined and at front, cursor becomes
//undefined, if cursor is undefined does nothing.
void movePrev(List L) {
    if( L==NULL ){
        printf("List Error: calling movePrev() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling movePrev() on an empty List\n");
        exit(1);
    }
    if( index(L) == -1){
        //do nothing
    }
    else {
        L->index--;
    }
}

//moveNext()
// If cursor is defined and not at back, moves cursor one step toward
// back of this List, if cursor is defined and at back, cursor becomes
// undefined, if cursor is undefined does nothing.
void moveNext(List L) {
    if( L==NULL ){
        printf("List Error: calling moveNext() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling moveNext() on an empty List\n");
        exit(1);
    }
    if( index(L) == -1) {
        //do nothing
    }
    else {
        L->index++;
        if( index(L)>= length(L)) {  //If the index >= length:
            L->index = -1;              //Cursor beomes undefined
        }
    }
}

// append()
// Places new data element at the end of L
void append(List L, int data)
{
    Node N = newNode(data);
    
    if( L==NULL ){
        printf("List Error: calling append() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ) {
        L->front = L->back = N;
    }else{
        L->back->next = N;
        N->prev = L->back;
        L->back = N;
    }
    L->length++;
}

// prepend()
// Insert new element into this List.
// If List is non-empty, insertion takes place before front element
void prepend(List L, int data) {
    Node N = newNode(data);
    
    if( L==NULL ){
        printf("List Error: calling prepend() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ) {
        L->front = L->back = N;
    }
    else {
        N->next = L->front;
        L->front->prev = N;
        L->front = N;
    }
    L->length++;
    L->index++;
}

//insertBefore()
//Insert new element before cursor
//Pre: length()>0, index()>=0
void insertBefore(List L, int data) {
    if( L==NULL ) {
        printf("List Error: calling insertBefore() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ) {
        printf("List Error: calling insertBefore() on an empty List\n");
        exit(1);
    }
    if( index(L) < 0) {
        printf("List Error: calling insertBefore() on an undefined cursor\n");
        exit(1);
    }
    else if( index(L) == 0) {                //If inserting before first element
        prepend(L, data);
    }
    else {
        Node N = newNode(data);
        Node temp = L->front;
        for(int i = 1; i <= index(L); i++) {
            temp = temp->next;
        }
        N->next = temp;
        N->prev = temp->prev;
        temp->prev->next = N;
        temp->prev = N;
        L->length++;
        L->index++;                     //maintain cursor position
    }
}

//insertAfter()
//Insert new element after cursor
//Pre: length()>0, index()>=0
void insertAfter(List L, int data) {
    if( L==NULL ) {
        printf("List Error: calling insertAfter() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ) {
        printf("List Error: calling insertAfter() on an empty List\n");
        exit(1);
    }
    if( index(L) < 0) {
        printf("List Error: calling insertAfter() on an undefined cursor\n");
        exit(1);
    }
    if( index(L) == (length(L) - 1)) {  //If inserting after last element
        append(L, data);
    }
    else {
        Node N = newNode(data);
        Node temp = L->front;
        for(int i = 1; i <= index(L); i++) {
            temp = temp->next;
        }
        temp->next->prev = N;
        N->next = temp->next;
        N->prev = temp;
        temp->next = N;
        L->length++;
    }
}

// deleteFront()
// Deletes element at front of L
// Pre: !isEmpty(L)
void deleteFront(List L){
    Node N = NULL;
    
    if( L==NULL ){
        printf("List Error: calling deleteFront() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling deleteFront() on an empty List\n");
        exit(1);
    }
    N = L->front;
    if( length(L)>1 ) {
        L->front = L->front->next;
        L->front->prev = NULL;
        if( index(L) == 0) {
            L->index = -1;
        }
        else {
            L->index--;
        }
    }else{
        L->front = L->back = NULL;
        L->index = -1;
    }
    L->length--;
    freeNode(&N);
}

//deleteBack()
//Deletes the back element
//Pre: !this.isEmpty()
void deleteBack(List L){
    Node N = NULL;
    
    if( L==NULL ){
        printf("List Error: calling deleteBack() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ){
        printf("List Error: calling deleteBack() on an empty List\n");
        exit(1);
    }
    N = L->back;
    if( length(L) > 1) {
        L->back = L->back->prev;
        L->back->next = NULL;
        if(index(L) == (length(L) - 1)) {
            L->index = -1;
        }
    }
    else {
        L->front = L->back = NULL;
        L->index = -1;
    }
    L->length--;
    freeNode(&N);
}

//delete()
//Deletes cursor element, making it undefined
//Pre: length()>0, index()>=0;
void delete(List L) {
    if( L==NULL ) {
        printf("List Error: calling delete() on NULL List reference\n");
        exit(1);
    }
    if( isEmpty(L) ) {
        printf("List Error: calling delete() on an empty List\n");
        exit(1);
    }
    if( index(L) < 0) {
        printf("List Error: calling delete() on an undefined cursor\n");
        exit(1);
    }
    else if( index(L) == 0) {        //If the cursor is pointing at the front
        deleteFront(L);
        L->index = -1;
    }
    else if( index(L) == (length(L) - 1)) {  //If cursor is pointing at back
        deleteBack(L);
        L->index = -1;
    }
    else {
        Node temp = L->front;
        for(int i = 1; i <= index(L); i++) {
            temp = temp->next;
        }
        temp->next->prev = temp->prev;
        temp->prev->next = temp->next;
        freeNode(&temp);
        L->length--;
        L->index = -1;
    }
}

// Other Functions ------------------------------------------------------------

// printList()
// Prints data elements in L on a single line to stdout.
void printList(FILE* out, List L){
    Node N = NULL;
    
    if( L==NULL ){
        printf("List Error: calling printList() on NULL List reference\n");
        exit(1);
    }
    
    for(N = L->front; N != NULL; N = N->next){
        fprintf(out, " %d", N->data);
    }
    //printf("\n");
}

//copyList()
//Returns a new List identical to this List
List copyList(List L) {
    List Q = newList();
    Node N = L->front;
    
    while(N != NULL) {
        append(Q, N->data);
        N = N->next;
    }
    return Q;
}

File: CMPS 101 - C:Java/pa1/List.java
//Nichole Boothroyd     CruzID: nboothro    Assignment: pa1
//-----------------------------------------------------------------------------
//  Queue.java
//  An integer queue ADT
//
//  This file constitutes the solution to the first excercise in the ADT handout.
//  Recall that all ADT operations must state their own preconditions in a comment
//  block, then check that those conditions are satisfied.  If a precondition is
//  violated, the ADT operation should cause the program to quit with a useful error
//  message.  See functions getFront() and Dequeue() below for examples on how to do
//  this. Observe that we've added two other operations called equals() and copy().
//
//
//
//      Modifying given Queue.java code to implement List module.
//-----------------------------------------------------------------------------
public class List{
    
    private class Node{
        // Fields
        int data;
        Node next;
        Node prev;
        
        // Constructor
        Node(int data) { this.data = data; next = null; prev = null; }
        
        // toString():  overrides Object's toString() method
        public String toString() {
            return String.valueOf(data);
        }
        
        // equals(): overrides Object's equals() method
        public boolean equals(Object x){
            boolean eq = false;
            Node that;
            if(x instanceof Node){
                that = (Node) x;
                eq = (this.data==that.data);
            }
            return eq;
        }
    }
    
    // Fields
    private Node front;
    private Node back;
    private int length;
    private int index;
    
    // Constructor:  Creates a new list [empty]
    List() {
        front = back = null;
        length = 0;
        index = -1;              //initialized at -1?  May need to come back to this. ..
    }
    
    
    // Access Functions --------------------------------------------------------
    
    // isEmpty()
    // Returns true if this List is empty, false otherwise.
    public boolean isEmpty() {
        return length==0;
    }
    
    // length()
    // Returns length [number of elements] of this List.
    public int length() {
        return length;
    }
    
    //index()
    //If the cursor is defined; returns the index of the cursor element, otherwise, returns -1
    public int index() {
        return index;       //index initialized at -1, will already return -1 if undefined
    }
    
    // front()
    // Returns front element.
    // Pre: !this.isEmpty()
    public int front(){
        if( this.isEmpty() ){
            throw new RuntimeException(
                                       "List Error: front() called on empty List");
        }
        return front.data;
    }
    
    //back()
    //Returns back element.
    //Pre: !this.isEmpty()
    public int back(){
        if( this.isEmpty() ){
            throw new RuntimeException(
                                       "List Error: back() called on an empty List");
        }
        return back.data;
    }
    
    //get()
    //Returns cursor element.
    //Pre: !this.isEmpty()
    //Pre: index needs to be defined
    public int get(){
        if( this.isEmpty() | this.index() < 0){
            throw new RuntimeException (
                                        "List Error: get() called on empty list or invalid index");
        }
        Node temp = front;
        for(int i = 1; i <= this.index(); i++) {
            temp = temp.next;
        }
        return temp.data;
    }
    
    // equals()
    // Overrides Object's equals() method.  Returns true if x is a List storing
    // the same integer sequence as this List, false otherwise.
    public boolean equals(List L){
        boolean eq  = false;
        List Q;
        Node N, M;
        
        if(L instanceof List){
            Q = (List)L;
            N = this.front;
            M = Q.front;
            eq = (this.length==Q.length);
            while( eq && N!=null ){
                eq = N.equals(M);
                N = N.next;
                M = M.next;
            }
        }
        return eq;
    }
    // Manipulation Procedures -------------------------------------------------
   
    //clear ()
    //Resets this List to its original empty state
    //Pre: !this.isEmpty()
    void clear() {
        if(this.isEmpty()) {
            throw new RuntimeException(
                                       "List Error: List is already empty");
        }
        else {
            front = back = null;
            length = 0;
            index = -1;
        }   //Should ^this^ be deleting from memory instead of just reassigning?
    }
    
    //moveFront()
    //Places the cursor under front element, or does nothing [not even an error message?]
    //Pre: !this.isEmpty()
    void moveFront(){
            if(!this.isEmpty()){
                index = 0;          //the index for the first element should be 0
            }
    }
    
    //moveBack()
    //If the List is non-empty, places the cursor under the back element, otherwise does nothing
    //[no error message?]
    //Pre: !this.isEmpty()
    void moveBack(){
        if(!this.isEmpty()){
            index = (this.length() - 1);        //The index for the last element == (length-1) of the list
        }
    }
    
    //movePrev ()
    //If cursor is defined and not at front, moves cursor one step toward
    //front of this List, if cursor is defined and at front, cursor becomes
    //undefined, if cursor is undefined does nothing.
    void movePrev() {
        if(this.index() == -1) {        //if the cursor is undefined exit the function
            //do nothing
        }
        else {
            index--;                    //whether or not the cursor is at the front[0], just index - 1
        }                               //cursor will become undefined at [-1]
    }
    
    //moveNext()
    // If cursor is defined and not at back, moves cursor one step toward
    // back of this List, if cursor is defined and at back, cursor becomes
    // undefined, if cursor is undefined does nothing.
    void moveNext() {
        if(this.index() == -1) {
         //do nothing
        }
        if(this.index() == ((this.length() - 1))) {        //if the cursor is at the back
            index = -1;                                   //becomes undefined
       }
       else {
           index++;
       }
    }
    
    // append()
    // Appends data to back of this List.
    void append(int data){
        Node N = new Node(data);
        if( this.isEmpty() ) {
            front = back = N;
        }
        else{
            back.next = N;
            N.prev = back;
            back = N;
        }
        length++;
    }
   //prepend()
   //Insert new element into this List. If List is non-empty, insertion takes place before front element
    void prepend(int data) {
        Node N = new Node(data);
        if(this.isEmpty()) {
            front = back = N;
        }
        else {
            N.next = front;     //N.next is now pointing to what was originally the front element
            front.prev = N;     //set front.prev to point to the new N
            front = N;          //Set N as the new front
        }
        length++;
    }
   
    //insertBefore()
    //Insert new element before cursor
    //Pre: length()>0, index()>=0
    void insertBefore(int data) {
        if(this.length() == 0 | this.index() < 0) {
            throw new RuntimeException (
                                        "List Error: insertBefore() called on empty list or bad cursor");
        }
        else {
            Node N = new Node(data);
            Node temp = front;
            for(int i = 1; i <= this.index(); i++) {   //Traverse to cursor. If cursor is at front [0]: skip
                temp = temp.next;
            }
            N.next = temp;                     //Clean up all the node pointers
            N.prev = temp.prev;
 //           temp.prev.next = N;
            temp.prev = N;
            length++;
 //           index++;                            //move index up one to maintain cursor
        }
    }
    
    //insertAfter()
    //Inserts new element after cursor.
    //Pre: length()> 0, index()>=0
    void insertAfter(int data) {
        if(this.length() == 0 | this.index() < 0) {
            throw new RuntimeException (
                                        "List Error: insertAfter() called on empty list or bad cursor");
        }
        else {
            Node N = new Node(data);
            Node temp = front;
            for(int i = 1; i <= this.index(); i++) {
                temp = temp.next;
            }
            N.next = temp.next;         //New node's next points to cursor's next element
            N.prev = temp;              //New node's prev points to cursor
//            temp.next.prev = N;         //The node after temp now points its prev to N
            temp.next = N;              //Cursor's next points to new node
            length++;
        }
    }
    
    // deleteFront()
    // Deletes front element from this List.
    // Pre: !this.isEmpty()
    void deleteFront(){
        if(this.isEmpty()){
            throw new RuntimeException(
                                       "List Error: deleteFront() called on empty List");
        }
        if(this.length() > 1){
            front = front.next;
            front.prev = null;
        }else{
            front = back = null;
        }
        length--;
    }
    
    //deleteBack()
    //Deletes the back element
    //Pre: !this.isEmpty()
    void deleteBack() {
        if(this.isEmpty()){
            throw new RuntimeException(
                                       "List Error: deleteBack() called on empty List");
        }
        if(this.length() > 1) {
            back = back.prev;
            back.next = null;
        }
        else {
            front = back = null;
        }
        length--;
    }
    
    //delete()
    //Deletes cursor element, making it undefined
    //Pre: length()>0, index()>=0;
    void delete() {
        if( this.isEmpty() |this.index() < 0) {
            throw new RuntimeException(
                                       "List Error: delete() called on empty List or bad cursor");
        }
        else {
            Node temp = front;
            for(int i = 1; i <= this.index(); i++) {
                temp = temp.next;
            }
            Node N = temp.prev;
            Node M = temp.next;
            N.next = temp.next;         //The node before the cursor now points to the node after the cursor
            M.prev = temp.prev;         //The node after the cursor now points to the node before the cursor
            index = -1;                 //Cursor now undefined
            length--;
        }
    }
    
    
    // Other Functions ---------------------------------------------------------
    
    // toString()
    // Overides Object's toString() method.
    public String toString(){
        StringBuffer sb = new StringBuffer();
        Node N = front;
        while(N!=null){
            //sb.append(" ");
            sb.append(N.toString());
            sb.append(" ");
            N = N.next;
        }
        return new String(sb);
    }
    
    // copy()
    // Returns a new List identical to this List.
    List copy(){
        List Q = new List();
        Node N = this.front;
        
        while( N!=null ){
            Q.append(N.data);
            N = N.next;
        }
        return Q;
    }
    
}
